cloud-controller-manager/do/metrics.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2020 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import "github.com/prometheus/client_golang/prometheus"
+
+type metrics struct {
+	host               string
+	apiRequestDuration *prometheus.HistogramVec
+	runLoopDuration    *prometheus.HistogramVec
+	reconcileDuration  *prometheus.HistogramVec
+}
+
+// create metrics
+var (
+	apiRequestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "firewall",
+			Name:      "http_request_duration_seconds",
+			Help:      "This is a histogram for tracking the duration of firewall API requests.",
+			Buckets:   prometheus.ExponentialBuckets(0.01, 2, 14),
+		},
+		[]string{"method", "code"},
+	)
+	runLoopDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "firewall",
+			Name:      "run_loop_duration_seconds",
+			Help:      "This is a histogram for tracking the duration of the run loop and whether it succeeds or not.",
+			Buckets:   prometheus.ExponentialBuckets(0.01, 2, 15),
```

Comment:
In one of Timo's comments on a previous iteration of this PR he mentioned

> What I think would be more interesting to have is the duration of a loop iteration (similar to how we have eventHandlerDuration). If we then also include a label indicating success/failure, then I believe we'd get a good picture of how our Run() loop is performing. The histogram buckets should then be configured such that we can cover up to the maximum run frequency (5 minutes IIRC).

So it sounds to me like he suggested the buckets should be able to capture to to a maximum of 5 minute duration here? This current exponential bucket only creates buckets up to 163 seconds:

```
>>> for each in range(1, 15):
...   print("{}".format(0.01*2**each))
...
0.02
0.04
0.08
0.16
0.32
0.64
1.28
2.56
5.12
10.24
20.48
40.96
81.92
163.84
```

But I'm probably misunderstanding something important.

Reply:
Taking another look at [how `firewallReconcileFrequency` is used in the `actualRun` method](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/9dd809f3ea45f8cfa09826f262bc98979cf90398/cloud-controller-manager/do/firewall_controller.go#L158) confirms my suspicion that we should this such that we have meaningful buckets throughout the range of 0-300 seconds. Again I would suggest using a linear distribution with bucket boundaries that align with any SLOs we might care about.

Reply:
This should have been 16 in order to hit `327.68`. But yeah I think linear distribution may make more sense in this case too. I didn't imagine those first 9 or 10 numbers to be too helpful either. Good suggestion.